### DISCUSSION TITLE
Matching Clip to Its Image

### DISCUSSION THREAD

**Discussion by sgrove82** [pymupdf/PyMuPDF#2925] [Looking for help]
Hello, 

Im attempting to create a version of layers for each drawn element of a pdf -- ie text, drawings, images.  So far I have a function that infers seqno for images by looping through index of bbox_log and matching 2 of 4 bbox coordinates  along with a tolerance and filtering on image type. Drawings and Text already have seqnos so that is how I filled in the gaps with Image. This so far is working fine. My problem is some images have clipping masks and Id like to extract that info and associate it with the correct image, Im inclined to use something similar, matching 2 of the bbox coordinates between images and clips but I realize clipping paths could have a higher probably that have bboxs that have no matching coordinates, ie all sides of the clip are smaller or bigger. This would break the logic Im following. Ideally I could use something like a seqno or some other key in dicts that could associate the two items instead of trying to infer. So I wanted to ask before I get a bit too far down the rabbit hole if there is an easier way to link these two elements either natively or another approach? Thanks!

**Comment by JorjMcKie** [comment]
Vector graphics extraction currently is the only place where we join the "standard" results with clipping information.
I am aware that every other object on a page may be subject to the same type of display limitation. But requests like yours are rare ...

In the end I believe you have no choice other than following the path you sketched out.
One thing of irritation:
The clipping path applicable to an object should / will **_precede_** that object. Clipping paths do not apply to objects preceding it.
You may want to have a look at the output of `mutool trace input.pdf > input.xml`. This is a log of things present in a page's appearance source (the `/Contents` in PDF, but other document types have analogous constructs).
You should find your image in there - indented accordingly such that applicable clippings (there may be several stacked ones) can be seen.
You may even want to consider using an XML package to decipher this information ...
